File: AndArch/third/src/main/java/com/py/third/MyVpAdapter.java
package com.py.third;

import android.content.Context;
import android.support.annotation.NonNull;
import android.support.v4.view.PagerAdapter;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.Toast;

import java.util.List;

public class MyVpAdapter extends PagerAdapter {
    private List<Integer> list;
    private Context context;

    public MyVpAdapter(Context context, List<Integer> list) {
        this.context = context;
        this.list = list;
    }

    @Override
    public int getCount() {
        return list.size();
    }

    @Override
    public boolean isViewFromObject(@NonNull View view, @NonNull Object object) {
        return view == object;
    }

    @NonNull
    @Override
    public Object instantiateItem(@NonNull ViewGroup container, final int position) {
        ImageView iv = new ImageView(context);
        iv.setImageResource(list.get(position));
        container.addView(iv);

        iv.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Toast.makeText(context, position + "", Toast.LENGTH_SHORT).show();
            }
        });

        return iv;
    }

    @Override
    public void destroyItem(@NonNull ViewGroup container, int position, @NonNull Object object) {
        container.removeView((View) object);
    }
}

File: AndArch/third/src/main/java/com/py/third/viewpager/MyVpAdapter.kt
package com.py.third.viewpager;

import android.content.Context
import android.support.v4.view.PagerAdapter
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.Toast

class MyVpAdapter : PagerAdapter {

    private val list: List<Int>
    private val context: Context //可变变量一般要注意使用安全操作符?


    constructor(context: Context, list: List<Int>) {
        this.context = context
        this.list = list
    }

    override fun getCount(): Int {
        return list.size
    }

    override fun isViewFromObject(view: View, p1: Any): Boolean {
        return view == p1
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val iv = ImageView(context)
        iv.setImageResource(list[position])
        container.addView(iv)
        iv.setOnClickListener(View.OnClickListener {
            return@OnClickListener Toast.makeText(context, position.toString(), Toast.LENGTH_SHORT).show()
        })

        return iv
    }

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        container.removeView(`object` as View)
    }

}